icelake/src/catalog/wrapper.rs

```diff
@@ -0,0 +1,226 @@
+//! This module contains catalog wrapper.
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use prometheus::histogram_opts;
+use prometheus::labels;
+use prometheus::opts;
+use prometheus::register_counter_with_registry;
+use prometheus::register_histogram_with_registry;
+use prometheus::Counter;
+use prometheus::Histogram;
+use prometheus::DEFAULT_BUCKETS;
+
+use crate::types::PartitionSpec;
+use crate::types::Schema;
+use crate::Namespace;
+use crate::Result;
+use crate::Table;
+use crate::TableIdentifier;
+
+use super::UpdateTable;
+use super::{Catalog, CatalogRef, IcebergContext};
+
+struct CatalogMetrics {
```

Comment:
Can we implement metrics at catalog trait level? Like what we does for OpenDAL.

```rust
trait Catalog {}

trait CatalogLayer: Catalog {}
```

So all our users can reuse the same layer across different catalogs.

Reply:
Current approach using a wrapper to include the metrics, so it works the same.

Reply:
> Current approach using a wrapper to include the metrics, so it works the same.

Wrapper seems not a good idea since it's not composable.

- Users have to install and compile `prometheus` even they don't want to use it.
- Users can't switch to other implementations like `prometheus-client`

I expect we can implement those logic as a middleware instead.

Reply:
It would be difficult to make it a layered architecture. `icelake` not  only expose catalog api, but also writers api. It would be huge effort if we want to make everything layered.

Reply:
> It would be difficult to make it a layered architecture.

The difficult will be moved to hiding `prometheus` under a feature. Current design makes it even harder since there are many locations to change.

> It would be difficult to make it a layered architecture. `icelake` not only expose catalog api, but also writers api. It would be huge effort if we want to make everything layered.

Perhaps we shouldn't perform metrics on the writers API since they are already covered by opendal.

Reply:
Then we do need to figure out a way for clean isolation. 😭 We still need the writer metrics since it would be useful for troubleshooting.